### PR TITLE
[Master] Fix query expression looping a stream should not contain () as its result type

### LINKED ISSUE #37121: Query expression looping a stream should not contain () as its result type
**Description:**
```ballerina
class NumberGeneratorWithError {
    int i = 0;
    public isolated function next() returns record {|int value;|}|error? {
        if (self.i == 3) {
            return error("Custom error thrown explicitly.");
        }
        self.i += 1;
        return {value: self.i};
    }
}

public function main() {
    NumberGeneratorWithError numGen = new;
    stream<int, error?> numberStream = new(numGen);

    //error, found int[]|error?
    int[]|error oddNumberList = from int num in numberStream
                                where (num % 2 == 1)
                                select num;
}
```

From the spec;
> for each input frame f
evaluate the in expression with f in scope resulting in an iterable value c;
create an Iterator object i from c
do the following in a loop L
> - call i.next() resulting in a value r;
> - if r is an error, then complete execution of the from-clause early with error r;
> - if r is (), stop loop L;

Ref: https://ballerina.io/spec/lang/2022R1/#from-clause

**Steps to reproduce:**

**Affected Versions:**
2201.1.1

**OS, DB, other environment details and versions:**

**Related Issues (optional):**

**Suggested Labels (optional):**

**Suggested Assignees (optional):**

### PR DESCRIPTION
**Pull request by prakanth97**
## Purpose

Fixes #37121

## Samples
```
class NumberGeneratorWithError {
    int i = 0;
    public isolated function next() returns record {|int value;|}|error? {
        if (self.i == 3) {
            return error("Custom error thrown explicitly.");
        }
        self.i += 1;
        return {value: self.i};
    }
}

public function main() {
    NumberGeneratorWithError numGen = new;
    stream<int, error?> numberStream = new(numGen);

    // Now this is not an error
    int[]|error oddNumberList = from int num in numberStream
                                where (num % 2 == 1)
                                select num;
}
```
type of oddNumberList is `int[]|error` not `int[]|error?`
## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [x] Added necessary tests
  - [x] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java [hunk 1]
```diff
@@ -6071,6 +6071,9 @@ private BType resolveQueryType(SymbolEnv env, BLangExpression selectExp, BType c
             }
 
             if (completionType != null && completionType.tag != TypeTags.NIL) {
+                if (completionType.tag == TypeTags.UNION) {
+                    ((BUnionType) completionType).getMemberTypes().remove(symTable.nilType);
```
Comment (pcnfernando) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java:6075]: use `Types.getSafeType()` instead

Comment (prakanth97) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java]: Updated 59d6e0f51dfc0da1d13def43bf1937727b0bc1a1

File: tests/jballerina-unit-test/src/test/resources/test-src/query/simple-query-with-defined-type.bal [hunk 2]
```diff
@@ -344,7 +344,7 @@ public function testQueryWithStream() returns boolean {
     NumberGenerator numGen = new;
     var numberStream = new stream<int, error?>(numGen);
 
-    int[]|error? oddNumberList = from int num in numberStream
+    int[]|error oddNumberList = from int num in numberStream
```
Comment (pcnfernando) [inline tests/jballerina-unit-test/src/test/resources/test-src/query/simple-query-with-defined-type.bal:347]: Shall we add few more tests that we use query-exp to loop through a stream. Try out different completion types of the stream. eg: stream<Type, error>, stream<Type>, stream<Type, error?>, stream<Type, Error1>, etc

Comment (prakanth97) [reply tests/jballerina-unit-test/src/test/resources/test-src/query/simple-query-with-defined-type.bal:371]: Updated a0d77b8c38b164221098dea81a8dabfacc3ef8e6

File: tests/jballerina-unit-test/src/test/resources/test-src/query/simple-query-with-defined-type.bal [hunk 3]
```diff
@@ -368,6 +392,29 @@ public function testQueryStreamWithError() {
     panic error("Expeted error, found: " + (typeof oddNumberList).toString());
 }
 
+public function testQueryStreamWithDifferentCompletionType() {
```
Comment (pcnfernando) [inline tests/jballerina-unit-test/src/test/resources/test-src/query/simple-query-with-defined-type.bal:395]:
```suggestion
public function testQueryStreamWithDifferentCompletionTypes() {
```

File: tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/query/SimpleQueryExpressionWithDefinedTypeTest.java [hunk 4]
```diff
@@ -239,6 +239,11 @@ public void testQueryStreamWithError() {
         BRunUtil.invoke(result, "testQueryStreamWithError");
     }
 
+    @Test(description = "Query a stream with different completion type")
+    public void testQueryStreamWithDifferentCompletionType() {
+        BRunUtil.invoke(result, "testQueryStreamWithDifferentCompletionType");
```
Comment (pcnfernando) [inline tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/query/SimpleQueryExpressionWithDefinedTypeTest.java:244]:
```suggestion
        BRunUtil.invoke(result, "testQueryStreamWithDifferentCompletionTypes");
```

File: tests/jballerina-unit-test/src/test/resources/test-src/query/simple-query-with-defined-type.bal [hunk 5]
```diff
@@ -368,6 +392,29 @@ public function testQueryStreamWithError() {
...
+    NumberGeneratorWithError numGen1 = new;
+    stream<int, error?> numberStream1 = new (numGen1);
+
+    NumberGeneratorWithError2 numGen2 = new;
+    stream<int, error> numberStream2 = new (numGen2);
+
+    NumberGeneratorWithCustomError numGen3 = new;
+    stream<int, ErrorR1?> numberStream3 = new (numGen3);
+
+    int[]|error oddNumberList1 = from int num in numberStream1
```
Comment (pcnfernando) [inline tests/jballerina-unit-test/src/test/resources/test-src/query/simple-query-with-defined-type.bal:405]: Do the assertions for the results as well

Comment (prakanth97) [reply tests/jballerina-unit-test/src/test/resources/test-src/query/simple-query-with-defined-type.bal:405]: Updated 03065346a7a3b551105a676a36c7cba7ad9450d2

**Review by pcnfernando — APPROVED** [review]

**Pull request merged**